### PR TITLE
Changes to Definition struct to allow section args

### LINKED ISSUE #2765: add arguments for definition sections

### PR DESCRIPTION
**Pull request by ikaneshiro**
**Description of the Pull Request (PR):**

This PR changes many of the fields of the Definition struct in order to encapsulate the idea of a section having additional args. This requires some changes throughout the build package in order to accomodate the change. This also changes the parser to parse those arguements properly into the Definition struct. Also fixes some comments

**This fixes or addresses the following GitHub issues:**

- Fixes #2765 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: pkg/build/types/definition.go [hunk 1]
```diff
@@ -23,7 +23,7 @@ type Definition struct {
 }
```
Comment (WestleyK) [inline pkg/build/types/definition.go:24]: Copyright for this file should be:

```
// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
```

File: pkg/build/types/parser/deffile.go [hunk 2]
```diff
@@ -131,33 +131,83 @@ func getSectionName(line string) string {
 	return lineSplit[0]
 }
```
Comment (WestleyK) [inline pkg/build/types/parser/deffile.go:133]: Copyright for this file should also be:

```
// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
```

File: pkg/build/types/parser/deffile.go [hunk 3]
```diff
@@ -169,7 +219,7 @@ func doSections(s *bufio.Scanner, d *types.Definition) error {
 			}
 		} else {
 			//this is a section
```
Comment (WestleyK) [inline pkg/build/types/parser/deffile.go]: I would add a space here, for consistency:

```suggestion
 			// this is a section
```

File: pkg/build/types/parser/deffile_test.go [hunk 4]
```diff
@@ -90,6 +90,7 @@ func TestParseDefinitionFile(t *testing.T) {
 		{"NoHeaderComments", "testdata_good/noheadercomments/noheadercomments", "testdata_good/noheadercomments/noheadercomments.json"},
 		{"NoHeaderWhiteSpace", "testdata_good/noheaderwhitespace/noheaderwhitespace", "testdata_good/noheaderwhitespace/noheaderwhitespace.json"},
 		{"MultipleScripts", "testdata_good/multiplescripts/multiplescripts", "testdata_good/multiplescripts/multiplescripts.json"},
+		{"SectionArgs", "testdata_good/sectionargs/sectionargs", "testdata_good/sectionargs/sectionargs.json"},
 	}
```
Comment (WestleyK) [inline pkg/build/types/parser/deffile_test.go:182]: Copyright for this file should also be:

```
// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
```

**Review by WestleyK — CHANGES_REQUESTED** [review]
Couple suggestions. Otherwise, looks great!

File: internal/pkg/runtime/engines/imgbuild/create_linux.go [hunk 5]
```diff
@@ -183,8 +184,14 @@ func (engine *EngineOperations) CreateContainer(pid int, rpcConn net.Conn) error
...
+		if f.Args == "" {
+			files = f
+		}
+	}
 	// iterate through filetransfers
-	for _, transfer := range engine.EngineConfig.Recipe.BuildData.Files {
+	for _, transfer := range files.Files {
 		// sanity
 		if transfer.Src == "" {
 			sylog.Warningf("Attempt to copy file with no name...")
```
Comment (WestleyK) [inline internal/pkg/runtime/engines/imgbuild/create_linux.go:197]: Is there more to this warning message? I usually use `...` when I want to add something later.

File: internal/pkg/runtime/engines/imgbuild/create_linux.go [hunk 6]
```diff
@@ -183,8 +184,14 @@ func (engine *EngineOperations) CreateContainer(pid int, rpcConn net.Conn) error
 }
```
Comment (WestleyK) [inline internal/pkg/runtime/engines/imgbuild/create_linux.go:185]: On line: `202`, or `209`: return error message should have no caps.

File: pkg/build/types/parser/deffile.go [hunk 7]
```diff
@@ -131,51 +131,101 @@ func getSectionName(line string) string {
...
 
-// parseTokenSection splits the token into maximum 2 strings separated by a newline,
-// and then inserts the section into the sections map
-//
-func parseTokenSection(tok string, sections map[string]string) error {
+// parseTokenSection into appropriate components to be placed into a types.Script struct
+func parseTokenSection(tok string, sections map[string]*types.Script, files *[]types.Files) error {
 	split := strings.SplitN(tok, "\n", 2)
 	if len(split) != 2 {
 		return fmt.Errorf("Section %v: Could not be split into section name and body", split[0])
```
Comment (WestleyK) [inline pkg/build/types/parser/deffile.go:138]: Return error should have no caps.

```suggestion
 		return fmt.Errorf("section %v: could not be split into section name and body", split[0])
```

File: pkg/build/types/parser/deffile.go [hunk 8]
```diff
@@ -131,51 +131,101 @@ func getSectionName(line string) string {
...
+		}
+
+		*files = append(*files, f)
+		return nil
+	}
+
 	if appSections[key] {
 		sectionSplit := strings.SplitN(strings.TrimLeft(split[0], "%"), " ", 3)
 		if len(sectionSplit) < 2 {
 			return fmt.Errorf("App Section %v: Could not be split into section name and app name", sectionSplit[0])
```
Comment (WestleyK) [inline pkg/build/types/parser/deffile.go:185]: Again, there should be no caps in return error.

```suggestion
 			return fmt.Errorf("app Section %v: could not be split into section name and app name", sectionSplit[0])
```

**Review by WestleyK — COMMENTED** [review]
Couple more suggestions...

**Comment by ikaneshiro** [comment]
Unnecessary as #3156 will include these changes.

**Pull request closed without merging**